src/PowerShell/Microsoft.WinGet.DSC/Microsoft.WinGet.DSC.psm1

```diff
@@ -193,130 +187,169 @@ class WinGetAdminSettings
-        if (($this.Action -eq [WinGetAction]::Full) -and ($this.Sources.Count -ne $currentSources.Count) -and ($this.Ensure -eq [Ensure]::Present))
+        $removeSource = $false
+        $resetSource = $false
+        $addSource = $false
+
+        if ($this.Ensure -eq [WinGetEnsure]::Present)
         {
-            return $false
+            if ($currentSource.Ensure -eq [WinGetEnsure]::Present)
+            {
+                if (-not $this.TestAgainstCurrent($currentSource))
+                {
+                    $resetSource = $true
+                    $addSource = $true
+                }
+                # else in desired state
+            }
+            else
+            {
+                $addSource = $true
+            }
+        }
+        else
+        {
+            if ($currentSource.Ensure -eq [WinGetEnsure]::Present)
+            {
+                $removeSource = $true
+            }
+            else
+            {
+                # In desired state of Absent
+            }
         }
 
-        foreach ($sourceName in $this.Sources.Keys)
+        if ($removeSource)
         {
-            #Check if the source name exists, if it doesn't, then return false.
-            $result = $currentSources.Keys | Where-Object { $_ -eq $sourceName }
-            if ($null -eq $result)
+            Remove-WinGetSource -Name $this.Name
+        }
+        # Only remove OR reset should be true, not both
+        elseif ($resetSource)
+        {
+            Reset-WinGetSource -Name $this.Name
+        }
+
+        if ($addSource)
+        {
+            $hashArgs = @{
+                Name = $this.Name
+                Argument = $this.Argument
+            }
+
+            if (-not [string]::IsNullOrWhiteSpace($this.Type))
             {
-                $currentState = [Ensure]::Absent
+                $hashArgs.Add("Type", $this.Type)
             }
 
-            # Type has a default value.
-            $source = $this.Sources.$($sourceName)
-            $sourceType = "Microsoft.PreIndexed.Package" # default source type
-            if ($source.ContainsKey("Type") -and (-not([string]::IsNullOrWhiteSpace($source.Type))))
+            if ($this.TrustLevel -ne [WinGetTrustLevel]::Undefined)
             {
-                $sourceType = $source.Type
+                $hashArgs.Add("TrustLevel", $this.TrustLevel)
             }
 
-            $existingSource = $currentSources.$($sourceName)
-            if ($source.Arg -ne $existingSource.Arg -or $sourceType -ne $existingSource.Type)
+            if ($null -ne $this.Explicit)
             {
-                $currentState = [Ensure]::Absent
+                $hashArgs.Add("Explicit", $this.Explicit)
             }
-        }
 
-        return $currentState -eq $this.Ensure
+            Add-WinGetSource @hashArgs
+        }
     }
-
-    # Sets the desired properties.
-    [void] Set()
+    
+    [bool] hidden TestAgainstCurrent([WinGetSource]$currentSource)

```

Comment:
I had to go back on why there's no comparison of Name. You don't want to call Get here because you don't want two Get-WinGetSource in Set. Maybe add a comment?